### DISCUSSION TITLE
[Validator] UnexpectedTypeException vs UnexpectedValueException

### DISCUSSION THREAD

**Discussion by alcohol** [symfony/symfony#53268] [Q&A]
What is the difference between these two and when should one or the other be used?

I noticed `UnexpectedTypeException` is always used if the constraint instanceof does not match, e.g.:

https://github.com/symfony/symfony/blob/f5d993bbbe0e4c68ab6b6a9d9b2ac2143a8f3732/src/Symfony/Component/Validator/Constraints/AtLeastOneOfValidator.php#L26

However, `UnexpectedValueException` and `UnexpectedTypeException` are used interchangeably in scenarios where the type of the `$value` argument is not what is expected. For example:

https://github.com/symfony/symfony/blob/f5d993bbbe0e4c68ab6b6a9d9b2ac2143a8f3732/src/Symfony/Component/Validator/Constraints/BicValidator.php#L77 vs https://github.com/symfony/symfony/blob/f5d993bbbe0e4c68ab6b6a9d9b2ac2143a8f3732/src/Symfony/Component/Validator/Constraints/JsonValidator.php#L34

Can anyone shed some light on this?

**Comment by MatTheCat** [comment]
See https://symfony.com/doc/current/validation/custom_constraint.html#creating-the-validator-itself:

> throw [UnexpectedValueException] if your validator cannot handle the passed type so that it can be marked as invalid

Indeed they will be converted to constraint violations: https://github.com/symfony/symfony/blob/8c5881e49ddc1ad3b6c7e7b4591e0e06596b06aa/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php#L751-L757

`UnexpectedTypeException` is more of a logic exception: it should never happen.

#27917 and #28477 were worked on simultaneously; that probably explains why the `JsonValidator` is not up-to-date.
